lib/hashie/extensions/strict_key_access.rb

```diff
@@ -0,0 +1,71 @@
+module Hashie
+  module Extensions
+    # SRP: This extension will fail an error whenever a key is accessed that does not exist in the hash.
+    #
+    #   EXAMPLE:
+    #
+    #     class StrictKeyAccessHash < Hash
+    #       include Hashie::Extensions::StrictKeyAccess
+    #     end
+    #
+    #     >> hash = StrictKeyAccessHash[foo: "bar"]
+    #     => {:foo=>"bar"}
+    #     >> hash[:foo]
+    #     => "bar"
+    #     >> hash[:cow]
+    #       KeyError: key not found: :cow
+    #
+    # NOTE: For googlers coming from Python to Ruby, this extension makes a Hash behave like a "Dictionary".
+    #
+    module StrictKeyAccess
+      class DefaultError < StandardError
+        def initialize(msg = 'Setting or using a default with Hashie::Extensions::StrictKeyAccess does not make sense', *args)
+          super
+        end
+      end
+
+      # NOTE: This extension would break the default behavior of Hash initialization:
+      #
+      #     >> a = StrictKeyAccessHash.new(a: :b)
+      #     => {}
+      #     >> a[:a]
+      #       KeyError: key not found: :a
+      #
+      # Includes the Hashie::Extensions::MergeInitializer extension to get around that problem.
+      # Also note that defaults still don't make any sense with a StrictKeyAccess.
+      def self.included(base)
+        # Can only include into classes with a hash initializer
+        base.send(:include, Hashie::Extensions::MergeInitializer)
```

Comment:
No worries! I think the whole "default value" implementation in Ruby is a little nonsensical. I think the behavior of `MergeInitializer` would make a lot more sense, but it would get rid of the simple example in the core docs:

(from http://ruby-doc.org/core-2.2.3/Hash.html#method-c-new)

```ruby
h = Hash.new("Go Fish")
h["a"] = 100
h["b"] = 200
h["a"]           #=> 100
h["c"]           #=> "Go Fish"
```

This confusing part is that you can then do this:

```ruby
h = Hash.new("Go Fish")
h["a"] = 100
h["b"] = 200
h["a"]           #=> 100
h["c"]           #=> "Go Fish"
# The following alters the single default object
h["c"].upcase!   #=> "GO FISH"
h["d"]           #=> "GO FISH"
```

Which is nonsense and confusing when you're not expecting it.